src/hotspot/share/gc/g1/g1CollectionSetCandidates.cpp

```diff
@@ -26,78 +26,282 @@
+  size_t reclaimed_from_marked_list = _marking_regions.remove(other);
+
+  for (HeapRegion* r : *other) {
+    assert(contains(r), "must contain region %u", r->hrm_index());
+    _contains_map[r->hrm_index()] = CandidateOrigin::Invalid;
+  }
+
+  assert(_reclaimable_bytes >= reclaimed_from_marked_list, "must be");
+  _reclaimable_bytes -= reclaimed_from_marked_list;
+
+  verify();
+}
+
+bool G1CollectionSetCandidates::is_empty() const {
+  return length() == 0;
+}
+
+bool G1CollectionSetCandidates::has_no_more_marking_candidates() const {
+  return _marking_regions.length() == 0;
 }
 
 #ifndef PRODUCT
-void G1CollectionSetCandidates::verify() const {
-  guarantee(_front_idx <= _num_regions, "Index: %u Num_regions: %u", _front_idx, _num_regions);
-  uint idx = _front_idx;
-  size_t sum_of_reclaimable_bytes = 0;
-  HeapRegion *prev = nullptr;
-  for (; idx < _num_regions; idx++) {
-    HeapRegion *cur = _regions[idx];
-    guarantee(cur != nullptr, "Regions after _front_idx %u cannot be NULL but %u is", _front_idx, idx);
-    // Currently the decision whether young gc moves region contents is determined
-    // at region allocation time. It is not possible that a region becomes non-movable
-    // at a later point, which means below condition always holds true.
-    guarantee(G1CollectionSetChooser::should_add(cur),
-              "Region %u should be eligible for addition.", cur->hrm_index());
-    if (prev != nullptr) {
-      guarantee(prev->gc_efficiency() >= cur->gc_efficiency(),
-                "GC efficiency for region %u: %1.4f smaller than for region %u: %1.4f",
-                prev->hrm_index(), prev->gc_efficiency(), cur->hrm_index(), cur->gc_efficiency());
+void G1CollectionSetCandidates::verify_helper(G1CollectionCandidateList* list, uint& from_marking, size_t& reclaimable_bytes, CandidateOrigin* verify_map) {
+  list->verify();
+
+  for (uint i = 0; i < (uint)list->length(); i++) {
+    HeapRegion* r = list->at(i)._r;
+
+    if (is_from_marking(r)) {
+      from_marking++;
     }
-    sum_of_reclaimable_bytes += cur->reclaimable_bytes();
-    prev = cur;
+    const uint hrm_index = r->hrm_index();
+    assert(_contains_map[hrm_index] == CandidateOrigin::Marking,
+           "must be %u is %u", hrm_index, (uint)_contains_map[hrm_index]);
+    assert(verify_map[hrm_index] == CandidateOrigin::Invalid, "already added");
+
+    verify_map[hrm_index] = CandidateOrigin::Verify;
+
+    reclaimable_bytes += r->reclaimable_bytes();
+  }
+}
+
+void G1CollectionSetCandidates::verify() {
+  uint from_marking = 0;
+  size_t reclaimable_bytes = 0;
+
+  CandidateOrigin* verify_map = NEW_C_HEAP_ARRAY(CandidateOrigin, _max_regions, mtGC);
+  for (uint i = 0; i < _max_regions; i++) {
+    verify_map[i] = CandidateOrigin::Invalid;
+  }
+
+  verify_helper(&_marking_regions, from_marking, reclaimable_bytes, verify_map);
+
+  assert(length() >= marking_regions_length(), "must be");
```

Comment:
Don't get what the intention is here, given `uint length() const { return marking_regions_length(); }`.